shared/utility/math/filter/ParticleFilter.hpp

```diff
@@ -38,242 +72,396 @@ namespace utility::math::filter {
-        std::mt19937 rng;
-        std::normal_distribution<Scalar> norm;
-
-        // Dimension types for vectors and square matricies
-        using ParticleList = Eigen::Matrix<Scalar, Model::size, Eigen::Dynamic>;
-
-        ParticleList sample_particles(const StateVec& mean, const StateMat& covariance, const int& n_particles) {
-            // Sample single gaussian (represented by a gaussian mixture model of size 1)
-
-            // Implementation based on the work presented in
-            // Conrad Sanderson and Ryan Curtin.
-            // An Open Source C++ Implementation of Multi-Threaded Gaussian Mixture Models, k-Means and
-            // Expectation Maximisation. International Conference on Signal Processing and Communication
-            // Systems, 2017.
-
-            ParticleList new_particles =
-                ParticleList::NullaryExpr(Model::size, n_particles, [&]() { return norm(rng); });
-
-            const StateMat sqrt_covariance = covariance.cwiseSqrt();
+        /**
+         * @brief Constructs the particle filter using default values. Mean is set to 0 and covariance is set to
+         * \f$0.1 * I_{n}\f$, where n is the size of the model. The residual method is used for resampling particles and
+         * the residual will use the systematic method for resampling the residual particles.
+         */
+        ParticleFilter() {
+            set_state(StateVec::Zero(), StateMat::Identity() * 0.1);
+            resample_method = ResampleMethod::RESIDUAL;
+            residual_method = ResampleMethod::SYSTEMATIC;
+        }
 
-            for (int i = 0; i < n_particles; ++i) {
-                new_particles.col(i) = sqrt_covariance * new_particles.col(i) + mean;
+        /**
+         * @brief Constructs the particle filter with the provided values. The resampling method with default to the
+         * residual method with the systematic method for resampling the residual particles. It is an error to use the
+         * residual method for resampling residuals and an exception will be thrown.
+         *
+         * @param mean The mean of the multivariate normal distribution. Used for sampling the initial particles in the
+         * filter.
+         * @param covariance The covariance of the multivariate normal distribution. Used for sampling the initial
+         * particles in the filter.
+         * @param resample_method_ The method to use when resampling particles during the time update step.
+         * @param residual_method_ When using ResampleMethod::RESIDUAL for the resample method, this method will be used
+         * when the residual method needs to resample residual particles.
+         */
+        ParticleFilter(const StateVec& mean,
+                       const StateMat& covariance,
+                       const ResampleMethod& resample_method_ = ResampleMethod::RESIDUAL,
+                       const ResampleMethod& residual_method_ = ResampleMethod::SYSTEMATIC) {
```

Comment:
Don't know how I feel about this one given that sometimes this parameter does nothing. Honestly it might make more sense to use this as a bitfield and just have a `flags` field that you set to `ResampleMethod::SYSTEMATIC | ResampleMethod::RESIDUAL`

Reply:
That also sounds kinda clunky tbh

Reply:
Yes, but less clunky until we find a better alternative